### PR TITLE
fix: zetaclient should pick up new connector and erc20 custody addresses

### PR DESCRIPTION
**Pull request by ws4charlie**
# Description

1. Zetaclient should be able to refresh connector/erc20Custody addresses at runtime.
2. Added unit tests.

Closes: [1633](https://github.com/zeta-chain/node/issues/1633)

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Include instructions and any relevant details so others can reproduce. 

- [ ] Tested CCTX in localnet
- [ ] Tested in development environment
- [x] Go unit tests
- [ ] Go integration tests
- [ ] Tested via GitHub Actions 

# Checklist:

- [ ] I have added unit tests that prove my fix feature works

### CODE REVIEW

File: zetaclient/evm/evm_signer.go [hunk 1]
```diff
@@ -33,20 +33,19 @@ import (
 )
 
 type Signer struct {
```
Comment (lumtis) [inline zetaclient/evm/evm_signer.go:35]: Can we add comment on this structure?

File: zetaclient/zetacore_observer.go [hunk 2]
```diff
@@ -317,15 +326,48 @@ func (co *CoreObserver) scheduleCctxBTC(
...
+				co.logger.ZetaChainWatcher.Info().Msgf(
+					"updated ERC20 custody address for chainID %d, new address: %s", chainID, erc20CustodyAddress)
+			}
+		}
+	}
+	return signer, nil
+}
+
+// GetUpdatedChainOb returns chain client with updated chain parameters
+func (co *CoreObserver) GetUpdatedChainOb(coreContext *corecontext.ZetaCoreContext, chainID int64) (interfaces.ChainClient, error) {
```
Comment (lumtis) [inline zetaclient/zetacore_observer.go:362]: Can we add unit test for this method?

**Comment by lumtis** [comment]
@ws4charlie now the client and signer for each chain update dynamically the chain params?

Can we confirm the following?
- If new chain support is setup but no activated yet `isSupported=false`, the ZetaClient will actively start supporting the chain when `isSupported` is changed to true in the chain param and we don't need to restart ZetaClient
- If we want to remove support for a chain, we can simply set `isSupported` to `false` in the chain param and ZetaClient will automatically stop interacting with the chain without the need to be restarted

**Comment by ws4charlie** [comment]
> @ws4charlie now the client and signer for each chain update dynamically the chain params?
> 
> Can we confirm the following?
> 
> * If new chain support is setup but no activated yet `isSupported=false`, the ZetaClient will actively start supporting the chain when `isSupported` is changed to true in the chain param and we don't need to restart ZetaClient
> * If we want to remove support for a chain, we can simply set `isSupported` to `false` in the chain param and ZetaClient will automatically stop interacting with the chain without the need to be restarted
   
1. Dynamically pause/unpause a connected chain is tracked by this issue: https://github.com/zeta-chain/node/issues/1883
2. Dynamically add a new chain is a manual process which will at least involve
   - Observers runners update their zetaclient config files and zetaclient should be able to take new config dynamically.
   - Zetaclient should be able to detect newly added chain with `IsSupported` flag enabled.
   - Zetaclient should also be able to dynamically create new ChainClient and ChainSigner objects and start new threads.

File: zetaclient/zetacore_observer.go [hunk 3]
```diff
@@ -135,23 +138,29 @@ func (co *CoreObserver) startCctxScheduler(appContext *appcontext.AppContext) {
...
+					supportedChains := coreContext.GetEnabledChains()
 					for _, c := range supportedChains {
 						if c.ChainId == co.bridge.ZetaChain().ChainId {
 							continue
 						}
-						signer := co.signerMap[c]
-
-						cctxList, totalPending, err := co.bridge.ListPendingCctx(c.ChainId)
+						// update chain parameters for signer and chain client
+						signer, err := co.GetUpdatedSigner(coreContext, c.ChainId)
```
[the remark this answers is not in the document]
Comment (ws4charlie) [reply zetaclient/zetacore_observer.go:148]: That's the idea case. The reason is that `Signer` doesn't really keep `ChainParams` in its memory. Only EVM signers need the connector and custody addresses, nothing else.

**Comment by lumtis** [comment]
>Dynamically add a new chain is a manual process which will at least involve
Observers runners update their zetaclient config files and zetaclient should be able to take new config file dynamically.
Zetaclient should be able to detect newly added chain with IsSupported flag enabled.
Zetaclient should also be able to dynamically create new ChainClient and ChainSigner objects and start new threads.

We should implement all of this

File: zetaclient/evm/evm_signer.go (hunk 1 above)
Comment (ws4charlie) [reply zetaclient/evm/evm_signer.go:36]: https://github.com/zeta-chain/node/pull/1890/commits/1321dccf1e8bc841a1d84e04ae3ea9140ad2ed74

File: zetaclient/evm/evm_signer.go [hunk 4]
```diff
@@ -96,6 +94,34 @@ func NewEVMSigner(
 	}, nil
 }
 
+// SetZetaConnectorAddress sets the zeta connector address
+func (signer *Signer) SetZetaConnectorAddress(addr ethcommon.Address) {
```
[the remark this answers is not in the document]
Comment (ws4charlie) [reply zetaclient/evm/evm_signer.go:99]: https://github.com/zeta-chain/node/pull/1890/commits/1321dccf1e8bc841a1d84e04ae3ea9140ad2ed74

File: zetaclient/zetacore_observer.go (hunk 2 above)
Comment (ws4charlie) [reply zetaclient/zetacore_observer.go]: https://github.com/zeta-chain/node/pull/1890/commits/1321dccf1e8bc841a1d84e04ae3ea9140ad2ed74

File: zetaclient/zetacore_observer.go [hunk 5]
```diff
@@ -317,15 +326,48 @@ func (co *CoreObserver) scheduleCctxBTC(
...
 }
 
-func (co *CoreObserver) getUpdatedChainOb(appContext *appcontext.AppContext, chainID int64) (interfaces.ChainClient, error) {
+// GetUpdatedSigner returns signer with updated chain parameters
+func (co *CoreObserver) GetUpdatedSigner(coreContext *corecontext.ZetaCoreContext, chainID int64) (interfaces.ChainSigner, error) {
+	signer, found := co.signerMap[chainID]
+	if !found {
+		return nil, fmt.Errorf("signer not found for chainID %d", chainID)
+	}
+	// update EVM signer parameters only
```
[the remark this answers is not in the document]
Comment (ws4charlie) [reply zetaclient/zetacore_observer.go]: https://github.com/zeta-chain/node/pull/1890/commits/1321dccf1e8bc841a1d84e04ae3ea9140ad2ed74

**Comment by ws4charlie** [comment]
> > Dynamically add a new chain is a manual process which will at least involve
> > Observers runners update their zetaclient config files and zetaclient should be able to take new config file dynamically.
> > Zetaclient should be able to detect newly added chain with IsSupported flag enabled.
> > Zetaclient should also be able to dynamically create new ChainClient and ChainSigner objects and start new threads.
> 
> We should implement all of this. Let's track issues if those don't exist

https://github.com/zeta-chain/node/issues/1904

File: zetaclient/testutils/constant.go [hunk 6]
```diff
@@ -6,12 +6,16 @@ const (
...
 	TSSAddressEVMAthens3 = "0x8531a5aB847ff5B22D855633C25ED1DA3255247e"
 	TSSAddressBTCAthens3 = "tb1qy9pqmk2pd9sv63g27jt8r657wy0d9ueeh0nqur"
+	TssPubkeyEVMAthens3  = "zetapub1addwnpepq28c57cvcs0a2htsem5zxr6qnlvq9mzhmm76z3jncsnzz32rclangr2g35p"
 
-	// some other address
-	OtherAddress = "0x21248Decd0B7EcB0F30186297766b8AB6496265b"
+	// some other addresses
+	OtherAddress1 = "0x21248Decd0B7EcB0F30186297766b8AB6496265b"
+	OtherAddress2 = "0x33A351C90aF486AebC35042Bb0544123cAed26AB"
+	OtherAddress3 = "0x86B77E4fBd07CFdCc486cAe4F2787fB5C5a62cd3"
```
Comment (lumtis) [inline zetaclient/testutils/constant.go:18]: We have the `testutilsample` package to get sample test addresses. `sample.EthAddress()`

We should use it.

**Review by lumtis — APPROVED** [review]
Looks good.

One comment, and need to address build failure

File: zetaclient/testutils/constant.go (hunk 6 above)
Comment (ws4charlie) [reply zetaclient/testutils/constant.go:18]: Used that in my other tests as well. I wanted to keep a few for better determinism in some tests.

**Review by kevinssgh — APPROVED** [review]
Looks good

**Pull request merged**